### PR TITLE
MSC locking scheme for shared memory RW was added.

### PR DESCRIPTION
**Pull request by Alexandr5Efimov**

### CODE REVIEW

File: arch/concurrency/locking/shmem_locking/test.c [hunk 1]
```diff
@@ -83,8 +83,10 @@ void parse_cmdline(int argc, char **argv)
 {
     int i;
     for(i=1; i < argc; i++){
+//        printf("Parsed option %s\n", argv[i]);
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:86]:
```suggestion

```

File: arch/concurrency/locking/shmem_locking/test.c [hunk 2]
```diff
@@ -83,8 +83,10 @@ void parse_cmdline(int argc, char **argv)
...
         if( !strcmp(argv[i], "-v") ){
             verbose = 1;
+//            printf("Verbose = %d \n", verbose);
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:89]:
```suggestion
```

File: arch/concurrency/locking/shmem_locking/test.c [hunk 3]
```diff
@@ -485,8 +487,9 @@ int main(int argc, char **argv)
     MPI_Comm_rank(MPI_COMM_WORLD, &rank);
     MPI_Comm_size(MPI_COMM_WORLD, &size);
     
-    parse_cmdline(argc, argv);
-
+    if (rank == 0){
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:490]: Why only rank = 0?

File: arch/concurrency/locking/shmem_locking/test.c [hunk 4]
```diff
@@ -502,6 +505,7 @@ int main(int argc, char **argv)
         data->counter1 = 0;
         data->counter2 = 0;
         shared_rwlock_create(&data->lock);
+	VERBOSE_OUT(verbose, "Shared Seg Created\n");
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:508]:
```suggestion
	  VERBOSE_OUT(verbose, "Shared Seg Created\n");
```

File: arch/concurrency/locking/shmem_locking/test.c [hunk 5]
```diff
@@ -516,21 +520,29 @@ int main(int argc, char **argv)
     init_time += GET_TS() - start;
 
     calibrate_sleep(data);
+    if (0 == rank){
+	VERBOSE_OUT(verbose, "Calibration Test Passed\n");
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:524]:
```suggestion
	  VERBOSE_OUT(verbose, "Calibration Test Passed\n");
```

File: arch/concurrency/locking/shmem_locking/test.c [hunk 6]
```diff
@@ -516,21 +520,29 @@ int main(int argc, char **argv)
...
+
 
     // Correctness verification
 #if (!defined (MY_PTHREAD_MUTEX) || ( MY_PTHREAD_MUTEX == 0 ))
     /* Mutexes will not pass this test as only one reader at the
      * time can get into the region - so the won't be able to
      * call barrier together
      */
-     verification(data);
+    verification(data);
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:534]:
```suggestion
     verification(data);
```

File: arch/concurrency/locking/shmem_locking/test.c [hunk 7]
```diff
@@ -516,21 +520,29 @@ int main(int argc, char **argv)
...
+    if (0 == rank)VERBOSE_OUT(verbose, "Verification Test Passed\n");
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:535]:
```suggestion
    if (0 == rank){
        VERBOSE_OUT(verbose, "Verification Test Passed\n");
    }
```

File: arch/concurrency/locking/shmem_locking/test.c [hunk 8]
```diff
@@ -516,21 +520,29 @@ int main(int argc, char **argv)
...
 #endif
     
 
     wronly_test(data, &wronly_lock_ovh);
+    if (0 == rank)    VERBOSE_OUT(verbose, "Write Only Test Passed\n");
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:540]: same fix

File: arch/concurrency/locking/shmem_locking/test.c [hunk 9]
```diff
@@ -516,21 +520,29 @@ int main(int argc, char **argv)
...
     rdonly_test(data, &rdonly_lock_ovh);
+    if (0 == rank)    VERBOSE_OUT(verbose, "Read Only Test Passed\n");
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:542]: same fix

File: arch/concurrency/locking/shmem_locking/test.c [hunk 10]
```diff
@@ -516,21 +520,29 @@ int main(int argc, char **argv)
...
     if( !nordwr ){
         rdwr_test(data, &rlock_cnt, &rdwr_wr_time, &rdwr_wr_lock_ovh);
+	if (0 == rank)	VERBOSE_OUT(verbose, "Read/Write Test Passed\n");
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:545]: same fix

File: arch/concurrency/locking/shmem_locking/1nmsc.c [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    if( sizeof(lock->locks) / sizeof(lock->locks[0]) * 2 < lock_num ) {
+        printf("Too many processes, cannot allocate enough locks\n");
+        MPI_Abort(MPI_COMM_WORLD, 0);
+    }
+
+    for(i = 0; i < lock_num; i++) {
+        lock->locks[i].e.head = 0;
+        lock->locks[i].e.record[0].next = 0;
+        lock->locks[i].e.record[1].next = 0;
+    }
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/1nmsc.c:37]:
```suggestion
    memset(lock->locks, 0, sizeof(lock->locks));
```

File: arch/concurrency/locking/shmem_locking/test.c (hunk 3 above)
Comment (Alexandr5Efimov) [reply arch/concurrency/locking/shmem_locking/test.c:489]: can't remember

File: arch/concurrency/locking/shmem_locking/test.c (hunk 6 above)
Comment (artpol84) [reply arch/concurrency/locking/shmem_locking/test.c:529]: this one is still a change while it's not a change

File: arch/concurrency/locking/shmem_locking/test.c [hunk 12]
```diff
@@ -516,21 +515,39 @@ int main(int argc, char **argv)
     init_time += GET_TS() - start;
 
     calibrate_sleep(data);
+    if (0 == rank){
+	VERBOSE_OUT(verbose, "Calibration Test Passed\n");
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:519]:
```suggestion
        VERBOSE_OUT(verbose, "Calibration Test Passed\n");
```

File: arch/concurrency/locking/shmem_locking/test.c (hunk 6 above)
Comment (artpol84) [reply arch/concurrency/locking/shmem_locking/test.c:529]:
```suggestion
    verification(data);
```

File: arch/concurrency/locking/shmem_locking/test.c [hunk 13]
```diff
@@ -516,21 +515,39 @@ int main(int argc, char **argv)
...
+
 
     // Correctness verification
 #if (!defined (MY_PTHREAD_MUTEX) || ( MY_PTHREAD_MUTEX == 0 ))
     /* Mutexes will not pass this test as only one reader at the
      * time can get into the region - so the won't be able to
      * call barrier together
      */
-     verification(data);
+    verification(data);
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:529]:
```suggestion
    verification(data);
```

File: arch/concurrency/locking/shmem_locking/x86.h [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+   unsigned char ret;
+   __asm__ __volatile__ (
+                       SMPLOCK "cmpxchgq %3,%2   \n\t"
+                               "sete     %0      \n\t"
+                       : "=qm" (ret), "+a" (oldval), "+m" (*addr)
+                       : "q"(newval)
+                       : "memory", "cc"
+                       );
+
+   return (int) ret;
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/x86.h:23]:
```suggestion
    unsigned char ret;
    __asm__ __volatile__ (
                       SMPLOCK "cmpxchgq %3,%2   \n\t"
                               "sete     %0      \n\t"
                       : "=qm" (ret), "+a" (oldval), "+m" (*addr)
                       : "q"(newval)
                       : "memory", "cc"
                       );

    return (int) ret;
```

File: arch/concurrency/locking/shmem_locking/test.c [hunk 15]
```diff
@@ -516,21 +515,39 @@ int main(int argc, char **argv)
...
     
 
     wronly_test(data, &wronly_lock_ovh);
+    if (0 == rank){
+	VERBOSE_OUT(verbose, "Write Only Test Passed\n");
+    }
+    
     rdonly_test(data, &rdonly_lock_ovh);
+    if (0 == rank){
+	VERBOSE_OUT(verbose, "Read Only Test Passed\n");
```
Comment (artpol84) [inline arch/concurrency/locking/shmem_locking/test.c:543]:
```suggestion
        VERBOSE_OUT(verbose, "Read Only Test Passed\n");
```

**Pull request closed without merging**